### PR TITLE
User account: Language Picker - Add grouping by territory

### PR DESCRIPTION
**Pull request by ramonjd**
This PR continues on from #13576.

We are grouping languages by territory in the Language Picker Modal. For a history of the discussion, see: #13576.

## Key features
1. Auto selection of language territory by user geolocation
2. Defaults to 'Popular' when no territory match found for user's geolocation
3. Languages that not connected to a Territory such as Esperanto and Interlingua are available via search.

with-popular-cat

## Testing
1. Go to `/me/account` and open the Language Picker

### Expectation
You should be able to browse and select WordPress.com-supported languages categorized by territory.

### CODE REVIEW

File: client/components/language-picker/modal.jsx [hunk 1]
```diff
@@ -3,22 +3,121 @@
...
+			'NI', 'PA', 'SV', 'AG', 'AI', 'AW', 'BB', 'BL', 'BQ', 'BS',
+			'CU', 'CW', 'DM', 'DO', 'GD', 'GP', 'HT', 'JM', 'KN', 'KY',
+			'LC', 'MF', 'MQ', 'MS', 'PR', 'SX', 'TC', 'TT', 'VC', 'VG',
+			'VI', 'AR', 'BO', 'BR', 'CL', 'CO', 'EC', 'FK', 'GF', 'GY',
+			'PE', 'PY', 'SR', 'UY', 'VE'
+		],
+	},
+	{
+		id: 'asia-pacific',
+		name: translate => translate( 'Asia-Pacific' ),
```
Comment (a8ci18n) [inline client/components/language-picker/modal.jsx]: Hi! I've found a possible matching string that has already been translated **95** times:
`translate( 'Pacific' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/language-picker/modal.jsx [hunk 2]
```diff
@@ -3,22 +3,121 @@
...
+			'CV', 'GH', 'GM', 'GN', 'GW', 'LR', 'ML', 'MR', 'NE', 'NG',
+			'SH', 'SL', 'SN', 'TG', 'AO', 'CD', 'CF', 'CG', 'CM', 'GA',
+			'GQ', 'ST', 'TD', 'BI', 'DJ', 'ER', 'ET', 'KE', 'KM', 'MG',
+			'MU', 'MW', 'MZ', 'RE', 'RW', 'SC', 'SO', 'SS', 'TZ', 'UG',
+			'YT', 'ZM', 'ZW', 'BW', 'LS', 'NA', 'SZ', 'ZA'
+		],
+	},
+	{
+		id: 'americas',
+		name: translate => translate( 'Americas' ),
```
Comment (a8ci18n) [inline client/components/language-picker/modal.jsx]: Hi! I've found a possible matching string that has already been translated **97** times:
`translate( 'America' )` *ES Score: 20*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/language-picker/constants.js [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+			'PY',
+			'SR',
+			'UY',
+			'VE',
+		],
+	},
+	{
+		id: 'asia-pacific',
+		default: true,
+		name: translate => translate( 'Asia-Pacific' ),
```
Comment (a8ci18n) [inline client/components/language-picker/constants.js]: Hi! I've found a possible matching string that has already been translated **95** times:
`translate( 'Pacific' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/language-picker/constants.js [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			'BW',
+			'LS',
+			'NA',
+			'SZ',
+			'ZA',
+		],
+	},
+	{
+		id: 'americas',
+		name: translate => translate( 'Americas' ),
```
Comment (a8ci18n) [inline client/components/language-picker/constants.js]: Hi! I've found a possible matching string that has already been translated **97** times:
`translate( 'America' )` *ES Score: 20*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by akirk** [comment]
Is it possible/easy to keep the Popular languages tab in addition to the regional ones?

**Comment by akirk** [comment]
This grouping is great, it makes the list much better digestible (and also interesting to see which languages are where), great that it handles languages available in multiple territories!

**Comment by folletto** [comment]
Looks great! :)

I'd +1 @akirk's suggestion above to keep the Popular ones which should be the ones we have most support. :)

**Comment by ramonjd** [comment]
Thanks again for the feedback.

Looks like Popular languages was originally there, but [removed for reasons](https://github.com/Automattic/wp-calypso/pull/13576#issuecomment-298882289).

Should be no sweat to put them back in.

**Comment by ramonjd** [comment]
> When selecting a non-Western European language, it doesn't switch to the right tab when opening the language picker after a page reload.

Given this, I'm questioning the utility of grabbing the user's geo location in order to display the default selected territory tab.

If, when opening the language picker, we should open the tab corresponding to the already-selected language (with the popular tab as a priority), then any other tab-selection logic becomes redundant.

I can't think of any another fallback scenario were the user's geo location would be useful.

**Comment by folletto** [comment]
> Looks like Popular languages was originally there, but removed for reasons.
> Should be no sweat to put them back in.

Yeah, I changed my mind, haven't i? :D 

The consideration I was working on right now is that while I would still not include "All", "Popular" might make sense as a way to show the ones that have the most coverage. That said, now that you point it out, it might not be very useful for a user perspective: even if these are the languages we cover better, it probably has little bearing on the ability to read them from the user...

Thoughts?

**Comment by ramonjd** [comment]
> The consideration I was working on right now is that while I would still not include "All", "Popular" might make sense as a way to show the ones that have the most coverage. That said, now that you point it out, it might not be very useful for a user perspective: 

A cursory view of the stats looks like we're covering (roughly) more than 80% of locales with the Popular tab, I'm just not sure how it adds to the UX. 

It's a hard one to gauge - I'll take your judgement on this one @folletto :)

**Comment by ramonjd** [comment]
How to reconcile these two:

> Geolocation + Browser metadata seem providing a better default than just showing the tab where the currently selected option is.

> When selecting a non-Western European language, it doesn't switch to the right tab when opening the language picker after a page reload.

Essentially, which tab do we show the user when they open the Language Picker?

1) The language group tab corresponding to their geo-location
2) The one in which their currently-selected language appears?
3) Popular?

How much it matters I can only guess. 

If I want to change to French, I'll proceed to find it regardless of which language group I'm shown first. But I can only speak for me! :) 

I tried looking for stats and, interestingly, we don't appear to track language switch events. 

I've requested access to ga stats, which we use to track general user settings changes so maybe that will assist.

**Comment by folletto** [comment]
> A cursory view of the stats looks like we're covering (roughly) more than 80% of locales with the Popular tab, I'm just not sure how it adds to the UX.

Given the one below (pre-selecting a tab) I'd say to keep it out then, and let's see how it goes.

Thanks for the reflection on this one. :) 


> The language group tab corresponding to their geo-location

I'd go for geo-location.

> How much it matters I can only guess.

I agree. We don't have anything right now, so I'd rather ship anything and then review once we know better. :)

> I tried looking for stats and, interestingly, we don't appear to track language switch events.

Can we add a tracking event? I expect to not be huge, but without analytics we're a bit blind. :)

**Comment by akirk** [comment]
> Can we add a tracking event? I expect to not be huge, but without analytics we're a bit blind. :)

For the record, we do track user locale changes but just on s (user_locale_change) in MC.

**Comment by ramonjd** [comment]
> For the record, we do track user locale changes but just with bump_stats (user_locale_change) in MC.

This is great to know. Actually, I didn't know how `bump_stats` worked until just now. Thank you!

I'm probably thinking too deeply into it, but I was mulling over going beyond just numbers to validate the geo-location tab approach, that is, how many users switch to a language on the tab that _we show them first_.

> We don't have anything right now, so I'd rather ship anything and then review once we know better. :)

Sounds good to me.

**Comment by ramonjd** [comment]
> I'd go for geo-location.

Done

**Comment by ramonjd** [comment]
Is this good to go @akirk ? See above for a discussion of which tab we show to the user. Currently it's geo.

**Pull request merged**